### PR TITLE
feat: job semaphore

### PR DESCRIPTION
**Pull request by crepererum**
Decouples the number of compaction jobs from the number of concurrent partitions. This is important once we have multiple compaction jobs per partition but also allows us to pre-fetch data while not being in the critical section (e.g. catalog data as it is done right not but also for features like #6650).

Also contains two small clean-ups (first two commits).

### CODE REVIEW

**Review by domodwyer — APPROVED** [review]
Are semaphores "fair" ? Would it be possible we starve a partition's compaction because it never gets the permit?

I don't think there is any need for ordered compaction tasks, but a fixed-size worker pool is another way of structuring this if we do.

**Comment by crepererum** [comment]
> Are semaphores "fair" ? Would it be possible we starve a partition's compaction because it never gets the permit?

Yes. It just uses the tokio semaphore under the hood. As per <https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html>:

> This Semaphore is fair, which means that permits are given out in the order they were requested. This fairness is also applied when acquire_many gets involved, so if a call to acquire_many at the front of the queue requests more permits than currently available, this can prevent a call to acquire from completing, even if the semaphore has enough permits complete the call to acquire.

So I don't think you can starve partitions with this.

**Comment by domodwyer** [comment]
Awesome - thanks for checking.

Fair semaphore makes sense as a default! I imagine people would get in all sorts of pickles otherwise.

File: compactor2/src/config.rs [hunk 1]
```diff
@@ -33,8 +33,15 @@ pub struct Config {
     pub backoff_config: BackoffConfig,
 
     /// Number of partitions that should be compacted in parallel.
+    ///
+    /// This should usually be larger than the compaction job concurrency since one partition can spawn multiple compaction jobs.
     pub partition_concurrency: NonZeroUsize,
 
+    /// Number of concurrent compaction jobs.
```
Comment (NGA-TRAN) [inline compactor2/src/config.rs:40]: make it `Number of concurrent compaction jobs per partition` will make it clearer

Comment (crepererum) [reply compactor2/src/config.rs:40]: They are NOT per partition. They are across partitions. This is exactly the point: no matter how deep / wide a partition branches in the divide&conquer system or how many partitions we currently have in preparation (e.g. perform catalog IO), the number of DF jobs (and hence the used memory/CPU) is limited to this number.

Comment (NGA-TRAN) [reply compactor2/src/config.rs:40]: Sounds great. Can't wait to see this running

**Pull request merged**